README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/ggmap)](https://cran.r-project.org/package=ggmap)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/dkahle/ggmap?branch=master&svg=true)](https://ci.appveyor.com/project/dkahle/ggmap)
<!-- badges: end -->

<hr>

# ggmap

**ggmap** is an R package that makes it easy to retrieve raster map
tiles from popular online mapping services like [Google
Maps](https://developers.google.com/maps/documentation/maps-static?hl=en),
[Stadia Maps](https://stadiamaps.com/), and
[OpenStreetMap](https://www.openstreetmap.org/), and plot them using the
[**ggplot2**](https://github.com/tidyverse/ggplot2) framework.

## Stadia Maps

Stadia Maps offers map tiles in several styles, including updated [tiles
from Stamen Design](https://stadiamaps.com/stamen/). An API key is
required, but no credit card is necessary to [sign
up](https://client.stadiamaps.com/signup) and there is a free tier for
non-commercial use. Once you have your API key, invoke the registration
function: `register_stadiamaps("YOUR-API-KEY", write = FALSE)`. Note
that setting `write = TRUE` will update your `~/.Renviron` file by
replacing/adding the relevant line. If you use the former, know that
you’ll need to re-do it every time you reset R.

Your API key is *private* and unique to you, so be careful not to share
it online, for example in a GitHub issue or saving it in a shared R
script file. If you share it inadvertently, just go to
client.stadiamaps.com, delete your API key, and create a new one.

``` r
library("ggmap")
#  Loading required package: ggplot2
#  ℹ Google's Terms of Service: <https://mapsplatform.google.com>
#    Stadia Maps' Terms of Service: <https://stadiamaps.com/terms-of-service/>
#    OpenStreetMap's Tile Usage Policy: <https://operations.osmfoundation.org/policies/tiles/>
#  ℹ Please cite ggmap if you use it! Use `citation("ggmap")` for details.

us <- c(left = -125, bottom = 25.75, right = -67, top = 49)
get_stadiamap(us, zoom = 5, maptype = "alidade_smooth") |> ggmap() 
#  ℹ © Stadia Maps © Stamen Design © OpenMapTiles © OpenStreetMap contributors.
```

![](tools/README-maptypes-1.png)

Use `qmplot()` in the same way you’d use `qplot()`, but with a map
automatically added in the background:

``` r
library("dplyr", warn.conflicts = FALSE)
library("forcats")

# define helper
`%notin%` <- function(lhs, rhs) !(lhs %in% rhs)

# reduce crime to violent crimes in downtown houston
violent_crimes <- crime |> 
  filter(
    offense %notin% c("auto theft", "theft", "burglary"),
    between(lon, -95.39681, -95.34188),
    between(lat, 29.73631, 29.78400)
  ) |> 
  mutate(
    offense = fct_drop(offense),
    offense = fct_relevel(offense, c("robbery", "aggravated assault", "rape", "murder"))
  )

# use qmplot to make a scatterplot on a map
qmplot(lon, lat, data = violent_crimes, maptype = "stamen_toner_lite", color = I("red"))
#  ℹ Using `zoom = 14`
#  ℹ © Stadia Maps © Stamen Design © OpenMapTiles © OpenStreetMap contributors.
```

![](tools/README-qmplot-1.png)

Often `qmplot()` is easiest because it automatically computes a nice
bounding box for you without having to pre-compute it for yourself, get
a map, and then use `ggmap(map)` in place of where you would ordinarily
(in a **ggplot2** formulation) use `ggplot()`. Nevertheless, doing it
yourself is more efficient. In that workflow you get the map first (and
you can visualize it with `ggmap()`):

``` r
bbox <- make_bbox(lon, lat, data = violent_crimes)
map <- get_stadiamap( bbox = bbox, maptype = "stamen_toner_lite", zoom = 14 )
#  ℹ © Stadia Maps © Stamen Design © OpenMapTiles © OpenStreetMap contributors.
ggmap(map)
```

![](tools/README-ggmap-1.png)

And then you layer on geoms/stats as you would with **ggplot2**. The
only difference is that (1) you need to specify the `data` arguments in
the layers and (2) the spatial aesthetics `x` and `y` are set to `lon`
and `lat`, respectively. (If they’re named something different in your
dataset, just put `mapping = aes(x = longitude, y = latitude))`, for
example.)

``` r
ggmap(map) +
  geom_point(data = violent_crimes, color = "red")
```

![](tools/README-ggmap-layers-1.png)

With **ggmap** you’re working with **ggplot2**, so you can add in other
kinds of layers, use
[**patchwork**](https://patchwork.data-imaginist.com), etc. All the
**ggplot2** geom’s are available. For example, you can make a contour
plot with `geom = "density2d"`:

``` r
library("patchwork")
library("ggdensity")

robberies <- violent_crimes |> filter(offense == "robbery")

points_map <- ggmap(map) + geom_point(data = robberies, color = "red")

# warnings disabled
hdr_map <- ggmap(map) + 
  geom_hdr(
    aes(lon, lat, fill = after_stat(probs)), data = robberies,
    alpha = .5
  ) +
  geomtextpath::geom_labeldensity2d(
    aes(lon, lat, level = after_stat(probs)),
    data = robberies, stat = "hdr_lines", size = 3, boxcolour = NA
  ) +
  scale_fill_brewer(palette = "YlOrRd") +
  theme(legend.position = "none")

(points_map + hdr_map) & 
  theme(axis.title = element_blank(), axis.text = element_blank(), axis.ticks = element_blank())
```

![](tools/README-ggmap-patchwork-1.png)

Faceting works, too:

``` r
ggmap(map, darken = .3) +
  geom_point(
    aes(lon, lat), data = violent_crimes, 
    shape = 21, color = "gray25", fill = "yellow"
  ) +
  facet_wrap(~ offense, nrow = 1) +
  theme(axis.title = element_blank(), axis.text = element_blank(), axis.ticks = element_blank())
```

![](tools/README-faceting-1.png)

## Google Maps

[Google Maps](https://cloud.google.com/maps-platform/terms) can be used
just as easily. However, since Google Maps use a center/zoom
specification, their input is a bit different:

``` r
(map <- get_googlemap("waco texas", zoom = 12))
#  ℹ <https://maps.googleapis.com/maps/api/staticmap?center=waco%20texas&zoom=12&size=640x640&scale=2&maptype=terrain&key=xxx>
#  ℹ <https://maps.googleapis.com/maps/api/geocode/json?address=waco+texas&key=xxx>
#  1280x1280 terrain map image from Google Maps; use `ggmap::ggmap()` to plot it.
ggmap(map)
```

![](tools/README-google_maps-1.png)

Moreover, you can get various different styles of Google Maps with
**ggmap** (just like Stadia Maps):

``` r
get_googlemap("waco texas", zoom = 12, maptype = "satellite") |> ggmap()
get_googlemap("waco texas", zoom = 12, maptype = "hybrid") |> ggmap()
get_googlemap("waco texas", zoom = 12, maptype = "roadmap") |> ggmap()
```

Google’s geocoding and reverse geocoding API’s are available through
`geocode()` and `revgeocode()`, respectively:

``` r
geocode("1301 S University Parks Dr, Waco, TX 76798")
#  ℹ <https://maps.googleapis.com/maps/api/geocode/json?address=1301+S+University+Parks+Dr,+Waco,+TX+76798&key=xxx>
#  # A tibble: 1 × 2
#      lon   lat
#    <dbl> <dbl>
#  1 -97.1  31.6
revgeocode(c(lon = -97.1161, lat = 31.55098))
#  ℹ <https://maps.googleapis.com/maps/api/geocode/json?latlng=31.55098,-97.1161&key=xxx>
#  Warning: Multiple addresses found, the first will be returned:
#  !   1301 S University Parks Dr, Waco, TX 76706, USA
#  !   55 Baylor Ave, Waco, TX 76706, USA
#  !   1437 S University Parks Dr, Waco, TX 76706, USA
#  !   HV2M+9H Waco, TX, USA
#  !   Bear Trail, Waco, TX 76706, USA
#  !   Robinson, TX 76706, USA
#  !   Waco, TX, USA
#  !   McLennan County, TX, USA
#  !   Texas, USA
#  !   United States
#  [1] "1301 S University Parks Dr, Waco, TX 76706, USA"
```

*Note: `geocode()` uses Google’s Geocoding API to geocode addresses.
Please take care not to disclose sensitive information. [Rundle, Bader,
and Moody (2022)](https://pmc.ncbi.nlm.nih.gov/articles/PMC8972108/)
have considered this issue and suggest various alternative options for
such data.*

There is also a `mutate_geocode()` that works similarly to
[**dplyr**](https://github.com/tidyverse/dplyr/)’s `mutate()` function:

``` r
tibble(address = c("white house", "", "waco texas")) |> 
  mutate_geocode(address)
#  ℹ <https://maps.googleapis.com/maps/api/geocode/json?address=white+house&key=xxx>
#  ℹ <https://maps.googleapis.com/maps/api/geocode/json?address=waco+texas&key=xxx>
#  # A tibble: 3 × 3
#    address         lon   lat
#    <chr>         <dbl> <dbl>
#  1 "white house" -77.0  38.9
#  2 ""             NA    NA  
#  3 "waco texas"  -97.1  31.6
```

Treks use Google’s routing API to give you routes (`route()` and
`trek()` give slightly different results; the latter hugs roads):

``` r
trek_df <- trek("houson, texas", "waco, texas", structure = "route")
#  ℹ <https://maps.googleapis.com/maps/api/directions/json?origin=houson,+texas&destination=waco,+texas&key=xxx&mode=driving&alternatives=false&units=metric>
qmap("college station, texas", zoom = 8) +
  geom_path(
    aes(x = lon, y = lat),  colour = "blue",
    size = 1.5, alpha = .5,
    data = trek_df, lineend = "round"
  )
#  ℹ <https://maps.googleapis.com/maps/api/staticmap?center=college%20station,%20texas&zoom=8&size=640x640&scale=2&maptype=terrain&language=en-EN&key=xxx>
#  ℹ <https://maps.googleapis.com/maps/api/geocode/json?address=college+station,+texas&key=xxx>
#  Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
#  ℹ Please use `linewidth` instead.
#  This warning is displayed once every 8 hours.
#  Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
#  generated.
```

![](tools/README-route_trek-1.png)

(They also provide information on how long it takes to get from point A
to point B.)

Map distances, in both length and anticipated time, can be computed with
`mapdist()`). Moreover the function is vectorized:

``` r
mapdist(c("houston, texas", "dallas"), "waco, texas")
#  ℹ <https://maps.googleapis.com/maps/api/distancematrix/json?origins=dallas&destinations=waco,+texas&key=xxx&mode=driving>
#  ℹ <https://maps.googleapis.com/maps/api/distancematrix/json?origins=houston,+texas&destinations=waco,+texas&key=xxx&mode=driving>
#  # A tibble: 2 × 9
#    from           to               m    km miles seconds minutes hours mode   
#    <chr>          <chr>        <int> <dbl> <dbl>   <int>   <dbl> <dbl> <chr>  
#  1 dallas         waco, texas 151451  151.  94.1    5004    83.4  1.39 driving
#  2 houston, texas waco, texas 295091  295. 183.    10054   168.   2.79 driving
```

## Google Maps API key

A few years ago Google has [changed its API
requirements](https://developers.google.com/maps/documentation/geocoding/usage-and-billing),
and **ggmap** users are now required to register with Google. From a
user’s perspective, there are essentially three ramifications of this:

1.  Users must register with Google. You can do this at
    <https://mapsplatform.google.com>. While it will require a valid
    credit card (sorry!), there seems to be a fair bit of free use
    before you incur charges, and even then the charges are modest for
    light use.

2.  Users must enable the APIs they intend to use. What may appear to
    **ggmap** users as one overarching “Google Maps” product, Google in
    fact has several services that it provides as geo-related solutions.
    For example, the [Maps Static
    API](https://developers.google.com/maps/documentation/maps-static/overview)
    provides map images, while the [Geocoding
    API](https://developers.google.com/maps/documentation/geocoding/overview)
    provides geocoding and reverse geocoding services. Apart from the
    relevant Terms of Service, generally **ggmap** users don’t need to
    think about the different services. For example, you just need to
    remember that `get_googlemap()` gets maps, `geocode()` geocodes
    (with Google, DSK is done), etc., and **ggmap** handles the queries
    for you. *However*, you do need to enable the APIs before you use
    them. You’ll only need to do that once, and then they’ll be ready
    for you to use. Enabling the APIs just means clicking a few radio
    buttons on the Google Maps Platform web interface listed above, so
    it’s easy.

3.  Inside R, after loading the new version of **ggmap**, you’ll need
    provide **ggmap** with your API key, a [hash
    value](https://en.wikipedia.org/wiki/Hash_function) (think string of
    jibberish) that authenticates you to Google’s servers. This can be
    done on a temporary basis with `register_google(key = "[your key]")`
    or permanently using
    `register_google(key = "[your key]", write = TRUE)` (note: this will
    overwrite your `~/.Renviron` file by replacing/adding the relevant
    line). If you use the former, know that you’ll need to re-do it
    every time you reset R.

Your API key is *private* and unique to you, so be careful not to share
it online, for example in a GitHub issue or saving it in a shared R
script file. If you share it inadvertantly, just get on Google’s website
and regenerate your key - this will retire the old one. Keeping your key
private is made a bit easier by **ggmap** scrubbing the key out of
queries by default, so when URLs are shown in your console, they’ll look
something like `key=xxx`. (Read the details section of the
`register_google()` documentation for a bit more info on this point.)

The new version of **ggmap** is now on CRAN soon, but you can install
the latest version, including an important bug fix in `mapdist()`, here
with:

``` r
if(!requireNamespace("devtools")) install.packages("devtools")
devtools::install_github("dkahle/ggmap")
```

## Installation

-   From CRAN: `install.packages("ggmap")`

-   From Github:

``` r
if (!requireNamespace("remotes")) install.packages("remotes")
remotes::install_github("dkahle/ggmap")
```


NEWS
ggmap 2.4
-------------------------------------------------------------------------

CHANGES

* ggmap now forms a directory in the temporary directory which caches queries.
  To persist the cache over multiple sessions, set option `ggmap.file_drawer`.
  See `file_drawer()` for more details.

* Functions related to the file draw and map caching are no longer exported:
  you shouldn't have been using these anyway!

* ggmap is now slightly less verbose: Terms of Service references are moved to package load.
* geocode has been given an overhaul; it now accepts a data argument for data frames (see examples), and no longer errors when it runs out of checks.
* stamen maps now hosts a wide new variety of maptypes that are kin of the existing ones.  see ?get_stamenmap.  some stamen maps have moved to jpgs; this has been fixed.
* the license is now GPL-2.
* cloudmade maps has been deprecated; it's no longer clear how long the service will be offered.


NEW FUNCTIONS

* calc_zoom: calc_zoom computes a reasonable zoom level for either (1) lon/lat variables in a data frame, (2) lon/lat ranges, or (3) a bounding box.
* get_navermap: use of the naver map service (thanks to Heewon Jeon).

NEW ARGUMENTS

* qmplot now accepts maptype.
* several functions now have a force argument that demands re-downloading of maps.
* get_googlemap now accepts an api_key argument for business users (thanks to Guy Carpenter for starting the process).
* geocode and revgeocode now accept client and signature arguments for business users.
* the style argument in get_googlemap is now much better (thanks @coxchristopher, @djhurio, and @mattmoehr).
* route now accepts transit for the mode argument (thanks @corynissan).

FIXES

* fixed a bug in stamen cloudmade maps that resulted in a very (very) small miscalculation in the bottom right bounding box of each tile.
* several spaces have been removed from google maps urls; in particular, language="zh-CN" now works properly  (thanks to @JiangXD.)
* unexported objects from ggplot2 and plyr are now properly handled.
* get_googlemap now properly creates a fileDrawer when needed (thanks Eric Anderson).
* authors attributions now use the person class.
* style argument fixed in get_googlemap (thanks @mattmoehr).
* a bug in get_stamenmap has been fixed that was making points at northern latitudes appear even moreso for very low zoom (thanks to @amyszczepanski's discovery).

ggmap 2.3
-------------------------------------------------------------------------

CHANGES

* now imports mapproj.
* theme_nothing now accepts argument legend to prevent legend wiping.
* ggmap and qmplot can now properly display legends in the map again, but transparency in the legend has been removed. (this will likely return if/when ggplot2 allows an alpha argument to element_rect).
* dataset crime now has variable offense as a factor, month and day as ordered factors.
* several functions now message the user with the URL queried as well as the Google Maps Terms of Service URL to better agree with Google's License Restrictions.
* gglocator now works properly.


DOCUMENTATION CHANGES

* documentation for get_map states that whole-world zooms are not supported. (thanks to suggestion by john hendy.)



ggmap 2.2
-------------------------------------------------------------------------

This version brings ggmap up to date with ggplot2 0.9.2.

CHANGES

* geocode now returns the query as well with the option output = "more".  this allows for easier merging with original datasets.  (thanks to suggestion by garrett grolemund.)
* geocode, mapdist, route and revgeocode now no longer close all connections. consequently, ggmap now works with knitr. (thanks to fix by yihui xie.)





ggmap 2.1
-------------------------------------------------------------------------

NEW FEATURES

* get_map now allows for bounding box specifications with source = 'google'.
* new argument darken added to ggmap which allows the user to put a shaded layer on top of maps.  this makes seeing color overlays easier.
* darken, mapcolor arguments added to qmplot.

CHANGES

* get_[google,osm,stamen,cloudmade] functions now check their arguments in the function themselves as opposed to using an alternative function for argument checking.
* ggmap arguments changed
	a. fullpage and expand collapsed into one argument called extent
	b. b changed to padding.  using b now issues a warning.
thanks to Jean-Olivier Irisson for the suggestions.
* deprecated syntaxes now properly call .Deprecated
* ggmap documentation greatly enhanced.
* get_stamenmap and get_cloudmademap now message the user if more than 20 tiles are needed.

FIXES

* get_map now attempts to determine an appropriate zoom when a bounding box is provided.
* qmplot now properly scales the expand on the x and y axes. (Thanks to Jean-Olivier Irisson for the suggestion.)


ggmap 2.0
-------------------------------------------------------------------------

NEW FEATURES

* a vast array of new kinds of maps are available to the user.  these are based on stamen maps (http://maps.stamen.com/watercolor/) and cloudmade maps (http://maps.cloudmade.com/).  the latter allows for user-designed map rendering (http://maps.cloudmade.com/editor#).

* the full google static maps api is available to the user.  in particular, this allows for double the image resolution of google maps (which is also now the default).

* qmplot - a version of qplot (from ggplot2) which automatically determines, downloads, and plots the result on the appropriate map.  this function is still experimental.

* make_bbox - a function which determines an appropriate bounding box for a collection of longitude/latitude pairs.

CHANGES

* ggmap now imports all packages save ggplot2 instead of depending on them.
* ggmapplot no longer uses a textConnection for parsing base_layer.
* geocode fixes/changes :
	a. gcdf is no longer a global variable
	b. geocode now properly uses the google geocoding api (like revgeocode)

R/qmplot.R
#' Quick map plot
#'
#' \code{qmplot} is the ggmap equivalent to the ggplot2 function qplot and allows for
#' the quick plotting of maps with data/models/etc.
#'
#'
#' @param x longitude values
#' @param y latitude values
#' @param ... other aesthetics passed for each layer
#' @param data data frame to use (optional).  If not specified, will create
#'   one, extracting vectors from the current environment.
#' @param zoom map zoom, see \code{\link{get_map}}
#' @param source map source, see \code{\link{get_map}}
#' @param maptype map type, see \code{\link{get_map}}
#' @param extent how much of the plot should the map take up? "normal", "panel", or "device" (default)
#' @param legend "left", "right" (default), "bottom", "top", "bottomleft", "bottomright", "topleft", "topright", "none" (used with extent = "device")
#' @param padding distance from legend to corner of the plot  (used with extent = "device")
#' @param force force new map (don't use archived version)
#' @param darken vector of the form c(number, color), where number is in [0, 1] and color is a character string indicating the color of the darken.  0 indicates no darkening, 1 indicates a black-out.
#' @param mapcolor color ("color") or black-and-white ("bw")
#' @param facets faceting formula to use.  Picks \code{\link{facet_wrap}} or
#'   \code{\link{facet_grid}} depending on whether the formula is one sided
#'   or two-sided
#' @param margins whether or not margins will be displayed
#' @param geom character vector specifying geom to use.  defaults to
#'  "point"
#' @param stat character vector specifying statistics to use
#' @param position character vector giving position adjustment to use
#' @param xlim limits for x axis
#' @param ylim limits for y axis
#' @param main character vector or expression for plot title
#' @param f number specifying the fraction by which the range should be extended
#' @param xlab character vector or expression for x axis label
#' @param ylab character vector or expression for y axis label
#' @export
#' @examples
#' \dontrun{
#' # these are skipped to conserve R check time
#'
#' qmplot(lon, lat, data = crime)
#'
#'
#' # only violent crimes
#' violent_crimes <- subset(crime,
#'   offense != "auto theft" &
#'   offense != "theft" &
#'   offense != "burglary"
#' )
#'
#' # rank violent crimes
#' violent_crimes$offense <-
#'   factor(violent_crimes$offense,
#'     levels = c("robbery", "aggravated assault",
#'       "rape", "murder")
#'   )
#'
#' # restrict to downtown
#' violent_crimes <- subset(violent_crimes,
#'   -95.39681 <= lon & lon <= -95.34188 &
#'    29.73631 <= lat & lat <=  29.78400
#' )
#'
#' theme_set(theme_bw())
#'
#' qmplot(lon, lat, data = violent_crimes, colour = offense, darken = .5,
#'   size = I(3.5), alpha = I(.6), legend = "topleft")
#'
#' qmplot(lon, lat, data = violent_crimes, geom = c("point","density2d"))
#' qmplot(lon, lat, data = violent_crimes) + facet_wrap(~ offense)
#' qmplot(lon, lat, data = violent_crimes, extent = "panel") + facet_wrap(~ offense)
#' qmplot(lon, lat, data = violent_crimes, extent = "panel", colour = offense,
#'     darken = .4) +
#'   facet_wrap(~ month)
#'
#'
#'
#'
#' # doesn't quite work yet.... http://tile.stamen.com/terrain/#4/30.28/-87.21
#' qmplot(long, lat, xend = long + delta_long,
#'   yend = lat + delta_lat, data = seals, geom = "segment")
#'
#' # this works though!
#' qmplot(long, lat, xend = long + delta_long, maptype = "toner-lite",
#'   color = I("red"), yend = lat + delta_lat, data = seals,
#'   geom = "segment", zoom = 5)
#'
#' qmplot(long, lat, xend = long + delta_long, maptype = "watercolor",
#'   yend = lat + delta_lat, data = seals,
#'   geom = "segment", zoom = 6)
#'
#'
#' library(scales)
#' library(grid)
#' qmplot(lon, lat, data = wind, size = I(.5), alpha = I(.5)) +
#'   ggtitle("NOAA Wind Report Sites")
#'
#' # thin down data set...
#' s <- seq(1, 227, 8)
#' thinwind <- subset(wind,
#'   lon %in% unique(wind$lon)[s] &
#'   lat %in% unique(wind$lat)[s]
#' )
#'
#' # for some reason adding arrows to the following plot bugs
#' theme_set(theme_bw(18))
#'
#' qmplot(lon, lat, data = thinwind, geom = "tile", fill = spd, alpha = spd,
#'     legend = "bottomleft") +
#'   geom_leg(aes(xend = lon + delta_lon, yend = lat + delta_lat)) +
#'   scale_fill_gradient2("Wind Speed\nand\nDirection",
#'     low = "green", mid = muted("green"), high = "red") +
#'   scale_alpha("Wind Speed\nand\nDirection", range = c(.1, .75)) +
#'   guides(fill = guide_legend(), alpha = guide_legend())
#'
#'
#'
#'
#' ## kriging
#' ############################################################
#' # the below examples show kriging based on undeclared packages
#' # to better comply with CRAN's standards, we remove it from
#' # executing, but leave the code as a kind of case-study
#' # they also require the rgdal library
#'
#' if(FALSE){
#'
#' library(lattice)
#' library(sp)
#' library(rgdal)
#'
#' # load in and format the meuse dataset (see bivand, pebesma, and gomez-rubio)
#' data(meuse)
#' coordinates(meuse) <- c("x", "y")
#' coordinates(meuse) <- ~x+y
#' proj4string(meuse) <- CRS("+init=epsg:28992")
#' meuse <- spTransform(meuse, CRS("+proj=longlat +datum=WGS84"))
#'
#' # plot
#' plot(meuse)
#'
#' df <- as.data.frame(slot(meuse, "coords"))
#' df <- data.frame(df, slot(meuse, "data"))
#'
#' qmplot(x, y, size = cadmium, alpha = I(.75), color = I("green"),
#'   data = df, zoom = 14, legend = "topleft",
#'   source = "google", maptype = "satellite", darken = .2
#' ) + scale_size("Cadmium\n(ppm)")
#'
#'
#' # load in the meuse.grid dataset (looking toward kriging)
#' library(gstat)
#' data(meuse.grid)
#' coordinates(meuse.grid) <- c("x", "y")
#' proj4string(meuse.grid) <- CRS("+init=epsg:28992")
#' meuse.grid <- spTransform(meuse.grid, CRS("+proj=longlat +datum=WGS84"))
#' mg <- as.data.frame(slot(meuse.grid, "coords"))
#' mg <- data.frame(mg, slot(meuse.grid, "data"))
#'
#' # plot it
#' qmplot(x, y, data = mg, source = "google")
#'
#'
#' # using linear regression - surface trend analysis
#' tsa <- krige(log(zinc) ~ 1, meuse, meuse.grid)
#' mg$pred <- slot(tsa, "data")$var1.pred
#' qmplot(x, y, data = mg, source = "google", color = pred,
#'     size = I(2.5), shape = I(15),
#'     alpha = I(3/5), darken = .3, legend = "topleft") +
#'   scale_color_gradient("Predicted\nLog Zinc",
#'     low = "green", high = "red"
#'   )
#'
#' tsa <- krige(log(zinc) ~ 1, meuse, meuse.grid, degree = 2)
#' mg$pred <- slot(tsa, "data")$var1.pred
#' qmplot(x, y, data = mg, source = "google", color = pred,
#'     size = I(2.5), shape = I(15),
#'     alpha = I(3/5), darken = .3, legend = "topleft") +
#'   scale_color_gradient("Predicted\nLog Zinc",
#'     low = "green", high = "red"
#'   )
#'
#'
#' # ordinary kriging
#' vgram <- variogram(log(zinc) ~ sqrt(dist), meuse)
#' vgramFit <- fit.variogram(vgram, vgm(1, "Exp", 300, 1))
#' ordinaryKrige <- krige(log(zinc) ~ 1, meuse, meuse.grid, vgramFit)
#' mg$pred <- slot(ordinaryKrige, "data")$var1.pred
#' qmplot(x, y, data = mg, source = "google", color = pred,
#'     size = I(2.5), shape = I(15),
#'     alpha = I(3/5), darken = .3, legend = "topleft") +
#'   scale_color_gradient("Predicted\nLog Zinc",
#'     low = "green", high = "red"
#'   )
#'
#' # universal kriging
#' universalKrige <- krige(log(zinc) ~ sqrt(dist), meuse, meuse.grid, vgramFit)
#' mg$pred <- slot(universalKrige, "data")$var1.pred
#' qmplot(x, y, data = mg, source = "google", color = pred,
#'     size = I(2.5), shape = I(15),
#'     alpha = I(3/4), darken = .3, legend = "topleft") +
#'   scale_color_gradient("Predicted\nLog Zinc",
#'     low = "green", high = "red"
#'   )
#'
#'
#'
#' # kriging with data
#' qmplot(x, y, data = mg, source = "google", color = pred,
#'     size = I(2.5), shape = I(15),
#'     alpha = I(3/4), darken = .3, legend = "topleft") +
#'   scale_color_gradient("Predicted\nLog Zinc",
#'     low = "green", high = "red"
#'   ) +
#'   geom_point(
#'     aes(x = x, y = y, size = log(zinc)), data = df,
#'     color = "black", alpha = 1/2
#'   ) +
#'   scale_size("Observed\nLog Zinc")
#'
#'
#'
#' } # end FALSE if
#'
#'
#'
#' } # end dontrun
#'
qmplot <- function(x, y, ..., data, zoom, source = "stamen", maptype = "terrain",
  extent = "device", legend = "right", padding = .02, force = FALSE,
  darken = c(0, "black"), mapcolor = "color",
  facets = NULL, margins = FALSE, geom = "auto", stat = list(NULL),
  position = list(NULL), xlim = c(NA, NA), ylim = c(NA, NA), main = NULL, f = 0.05,
  xlab = deparse(substitute(x)), ylab = deparse(substitute(y)))
{

  argnames <- names(as.list(match.call(expand.dots=FALSE)[-1]))
  arguments <- as.list(match.call()[-1])

  args <- as.list(match.call(expand.dots = TRUE)[-1])
  argsgiven <- names(args)
  if("mapcolour" %in% argsgiven) mapcolor <- eval(args$mapcolour)

  .all_aesthetics <- unlist(getAnywhere(.all_aesthetics)[1:42])
  aesthetics <- compact(arguments[.all_aesthetics])
  aesthetics <- aesthetics[!is.constant(aesthetics)]
  aes_names <- names(aesthetics)
  aesthetics <- rename_aes(aesthetics)
  class(aesthetics) <- "uneval"

  if (missing(data)) {
    # If data not explicitly specified, will be pulled from workspace
    data <- data.frame()

    # Faceting variables must be in a data frame, so pull those out
    facetvars <- all.vars(facets)
    facetvars <- facetvars[facetvars != "."]
    names(facetvars) <- facetvars
    facetsdf <- as.data.frame(lapply(facetvars, get))
    if (nrow(facetsdf)) data <- facetsdf
  }

  # Work out plot data, and modify aesthetics, if necessary
  if ("auto" %in% geom) {
    if (stat == "qq" || "sample" %in% aes_names) {
      geom[geom == "auto"] <- "point"
      stat <- "qq"
    } else if (missing(y)) {
      stop("y must be provided for quickmap.", call. = FALSE)
    } else {
      if (missing(x)) {
        aesthetics$x <- bquote(seq_along(.(y)), aesthetics)
      }
      geom[geom == "auto"] <- "point"
    }
  }

  env <- parent.frame()

  # calculate map dimensions
  bbox <- make_bbox(
    lon = data[,deparse(substitute(x))],
    lat = data[,deparse(substitute(y))],
    f = f
  )
  lon_range <- bbox[c("left","right")]
  lat_range <- bbox[c("bottom","top")]


  # compute zoom
  if(missing(zoom)){
    zoom <- calc_zoom(lon_range, lat_range)
    message(paste0("Using zoom = ", zoom, "..."))
  }


  # get map
  if(source == "google"){
    map <- get_map(location = c(mean(lon_range), mean(lat_range)), zoom = zoom,
      source = source, color = mapcolor,
      maptype = maptype, force = force
    )
  } else { # bounding box
    map <- get_map(location = bbox, zoom = zoom,
      source = source, color = mapcolor,
      maptype = maptype, force = force
    )
  }
  xmin <- attr(map, "bb")$ll.lon
  xmax <- attr(map, "bb")$ur.lon
  ymin <- attr(map, "bb")$ll.lat
  ymax <- attr(map, "bb")$ur.lat


  # check darken
  stopifnot(0 <= as.numeric(darken[1]) && as.numeric(darken[1]) <= 1)
  if(length(darken) == 1 & is.numeric(darken)) darken <- c(darken, "black")


  # initialize plot
  p <- ggplot(data, aesthetics, environment = env) +
    inset_raster(map, xmin, xmax, ymin, ymax) +
    annotate("rect", xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax,
  	  fill = darken[2], alpha = as.numeric(darken[1])) +
    coord_map(projection = "mercator")


  # enforce extent
  if(extent == "normal"){
    # nothing
  } else if(extent == "panel"){
  	p <- p +
      scale_x_continuous(lim = c(xmin, xmax), expand = c(0,0)) +
      scale_y_continuous(lim = c(ymin, ymax), expand = c(0,0))
  } else if(extent == "device"){
  	p <- p +
      scale_x_continuous(lim = c(xmin, xmax), expand = c(0,0)) +
      scale_y_continuous(lim = c(ymin, ymax), expand = c(0,0)) +
      theme_nothing(legend = TRUE)


    # legend for full device map
    if(legend %in% c("topleft","topright","bottomleft","bottomright")){
      if(legend == "bottomleft"){
        lp <- c(padding, padding)
        lj <- c(0,0)
      } else if(legend == "topleft"){
        lp <- c(padding, 1-padding)
        lj <- c(0,1)
      } else if(legend == "bottomright"){
        lp <- c(1-padding, padding)
        lj <- c(1,0)
      } else if(legend == "topright"){
        lp <- c(1-padding, 1-padding)
        lj <- c(1,1)
      }
      p <- p + theme(
        legend.position = lp, legend.justification = lj,
        legend.background = element_rect(
          fill = "white", colour = "gray80", size = .2
        )
      )
    } else if(legend %in% c("left","right","bottom","top")){
      p <- p + theme(legend.position = legend)
    } # else legend = "none" as part of theme_nothing()
  }



  if (is.null(facets)) {
    p <- p + facet_null()
  } else if (is.formula(facets) && length(facets) == 2) {
    p <- p + facet_wrap(facets)
  } else {
    p <- p + facet_grid(facets = deparse(facets), margins = margins)
  }

  if (!is.null(main)) p <- p + theme("title" = main)

  # Add geoms/statistics
  if (is.proto(position)) position <- list(position)

  mapply(function(g, s, ps) {

    if(is.character(g)) g <- Geom$find(g)
    if(is.character(s)) s <- Stat$find(s)
    if(is.character(ps)) ps <- Position$find(ps)

    params <- arguments[setdiff(names(arguments), c(aes_names, argnames))]
    params <- lapply(params, eval, parent.frame(n=1))

    p <<- p + layer(geom=g, stat=s, geom_params=params, stat_params=params, position=ps)
  }, geom, stat, position)

  if (!missing(xlab)) p <- p + xlab(xlab)
  if (!missing(ylab)) p <- p + ylab(ylab)

  if (!missing(xlim)) p <- p + xlim(xlim)
  if (!missing(ylim)) p <- p + ylim(ylim)



  p
}
